cli/src/complete.rs

```diff
@@ -44,6 +46,112 @@ pub fn local_bookmarks() -> Vec<CompletionCandidate> {
+            .collect())
+    })
+}
+
+pub fn untracked_bookmarks() -> Vec<CompletionCandidate> {
+    with_jj(|mut jj, config| {
+        let output = jj
+            .arg("bookmark")
+            .arg("list")
+            .arg("--all-remotes")
+            .arg("--template")
+            .arg(r#"if(remote && !tracked, name ++ "@" ++ remote ++ "\n")"#)
+            .output()
+            .map_err(user_error)?;
+
+        Ok(String::from_utf8_lossy(&output.stdout)
+            .lines()
+            .filter(|bookmark| !bookmark.ends_with("@git"))
+            .map(|bookmark| {
+                let mut display_order = 0;
+                if let Ok(prefix) = config.get::<String>("git.push-bookmark-prefix") {
+                    if !bookmark.starts_with(&prefix) {
+                        // other people's bookmarks are less interesting
+                        display_order += 1;
+                    }
+                }
+                CompletionCandidate::new(bookmark).display_order(Some(display_order))
+            })
+            .collect())
+    })
+}
+
+pub fn bookmarks() -> Vec<CompletionCandidate> {
+    with_jj(|mut jj, config| {
+        let output = jj
+            .arg("bookmark")
+            .arg("list")
+            .arg("--all-remotes")
+            .arg("--template")
+            .arg(r#"separate("@", name, remote) ++ "\n""#)
+            .output()
+            .map_err(user_error)?;
+        let stdout = String::from_utf8_lossy(&output.stdout);
+
+        let prefix = config.get::<String>("git.push-bookmark-prefix").ok();
+        let mut bookmarks = HashMap::new();
+
+        // To determine the importance of each bookmark, we figure out if it is
+        // tracked locally and if its push prefix matches our configured one.
+        for line in stdout.lines() {
+            let (bookmark, _) = line.split_once('@').unwrap_or((line, ""));
+            let properties = bookmarks.entry(bookmark).or_default();
+            let (mut local, mut mine) = *properties;
+
+            if !line.contains('@') {
+                local = true;
+            }
+            if let Some(prefix) = prefix.as_ref() {
+                if bookmark.starts_with(prefix) {
+                    mine = true;
+                }
+            }
+            *properties = (local, mine);
+        }
+
+        // Collect all bookmarks in four buckets, one for each combination of
+        // importance properties. The index of the bucket corresponds to its
+        // display order (0 is highest priority).
+        let mut ordered_bookmarks = vec![vec![]; 4];
+        for (bookmark, (local, mine)) in bookmarks.into_iter() {
+            match (local, mine) {
+                (true, true) => ordered_bookmarks[0].push(bookmark),
+                (true, false) => ordered_bookmarks[1].push(bookmark),
+                (false, true) => ordered_bookmarks[2].push(bookmark),
+                (false, false) => ordered_bookmarks[3].push(bookmark),
+            }
+        }
```

Comment:
nit: Perhaps, we can use `chunk_by()` to group bookmarks by name, and map to `(name, weight)`.
https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.chunk_by

Reply:
I thought about something like this, but I was unsure if it's absolutely guaranteed that `jj bookmark list` will return them grouped like that? I guess not grouping them would be considered worse UX, so it probably won't happen.

Reply:
It's unlikely we'll change the output order because the default `jj bookmark list` output relies on that (local name followed by remotes). Also, other completion functions assume that the output is sorted by name.